crates/symbolicator/src/services/download/mod.rs

```diff
@@ -199,23 +200,142 @@ async fn download_stream(
+}
+
+impl<'a> MeasureSourceDownloadGuard<'a> {
+    /// Creates a new measure guard for downloading a source.
+    pub fn new(task_name: &'a str, source_name: &'a str) -> Self {
+        Self {
+            state: MeasureState::Pending,
+            task_name,
+            source_name,
+            bytes_transferred: None,
+            creation_time: Instant::now(),
+        }
+    }
+
+    /// A checked add to the amount of bytes transferred during the download.
+    ///
+    /// This value will be emitted when the download's future is completed or cancelled.
+    pub fn add_bytes_transferred(&mut self, additional_bytes: u64) {
+        let bytes = self.bytes_transferred.get_or_insert(Default::default());
+        *bytes = bytes.saturating_add(additional_bytes);
+    }
+
+    /// Marks the download as terminated.
+    pub fn done<T, E>(mut self, reason: &Result<T, E>) {
+        self.state = MeasureState::Done(m::result(reason));
+    }
+}
+
+impl Drop for MeasureSourceDownloadGuard<'_> {
+    fn drop(&mut self) {
+        let status = match self.state {
+            MeasureState::Pending => "canceled",
+            MeasureState::Done(status) => status,
+        };
+
+        let duration = self.creation_time.elapsed();
+        let metric_name = format!("{}.duration", self.task_name);
+        metric!(
+            timer(&metric_name) = duration,
+            "status" => status,
+            "source" => self.source_name,
+        );
+
+        if let Some(bytes_transferred) = self.bytes_transferred {
+            // Times are recorded in milliseconds, so match that unit when calculating throughput,
+            // recording a byte / ms value.
+            // This falls back to the throughput being equivalent to the amount of bytes transferred
+            // if the duration is zero, or there are any conversion errors.
+            let throughput = (bytes_transferred as u128)
+                .checked_div(duration.as_millis())
+                .and_then(|t| t.try_into().ok())
+                .unwrap_or(bytes_transferred);
+            let throughput_name = format!("{}.throughput", self.task_name);
+            metric!(
+                histogram(&throughput_name) = throughput,
+                "status" => status,
+                "source" => self.source_name,
+            );
+        }
+    }
+}
+
+/// Measures the timing of a download-related future and reports metrics as a histogram.
+///
+/// This function reports a single metric corresponding to the task name. This metric is reported
+/// regardless of the future's return value.
+///
+/// A tag with the source name is also added to the metric, in addition to a tag recording the
+/// status of the future.
+fn measure_download_time<'a, F, T, E>(
+    source_name: &'a str,
+    f: F,
+) -> impl Future<Output = F::Output> + 'a
+where
+    F: 'a + Future<Output = Result<T, E>>,
+{
+    let guard = MeasureSourceDownloadGuard::new("source.download", source_name);
```

Comment:
Maybe name this `source.connect`? because `download = connect + stream` :-D

Reply:
Oh! Good point, it is done. What do you think about adding a common segment for those two as well so it's easy to know that they're related? ie `source.download.connect` and `source.download.stream`?

Reply:
Actually I'll make that change and we can update the tag later if there are any issues with it